daemon/regmaild/regmaild.c

```diff
@@ -270,6 +276,202 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
+
+    sqlite3 *Db = g_Db;
+    sqlite3_stmt *Stmt = NULL;
+
+    if (!req->vkey[0]) {
+        if (sqlite3_prepare(Db, "DELETE FROM verifydb "
+                    "WHERE userid = lower(?) AND generation = ? AND vmethod = ?;",
+                    -1, &Stmt, NULL) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_text(Stmt, 1, req->userid, -1, SQLITE_STATIC) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_int64(Stmt, 2, req->generation) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_int64(Stmt, 3, req->vmethod) != SQLITE_OK)
+            goto end;
+    } else {
+        if (sqlite3_prepare(Db, "REPLACE INTO verifydb "
+                    "(userid, generation, vmethod, vkey, timestamp) VALUES "
+                    "(lower(?),?,?,?,?);",
+                    -1, &Stmt, NULL) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_text(Stmt, 1, req->userid, -1, SQLITE_STATIC) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_int64(Stmt, 2, req->generation) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_int64(Stmt, 3, req->vmethod) != SQLITE_OK)
+            goto end;
+        if (sqlite3_bind_text(Stmt, 4, req->vkey, -1, SQLITE_STATIC) != SQLITE_OK)
+            goto end;
+        int64_t timestamp = time(NULL);
+        if (sqlite3_bind_int64(Stmt, 5, timestamp) != SQLITE_OK)
+            goto end;
+    }
+
+    if (sqlite3_step(Stmt) == SQLITE_DONE)
+        ret = VERIFYDB_OK;
+
+end:
+    if (Stmt != NULL)
+        sqlite3_finalize(Stmt);
+
+    return ret;
+}
+
+static void
+verifydb_get(const verifydb_req *req, int fd)
+{
+    sqlite3 *Db = g_Db;
+    sqlite3_stmt *Stmt = NULL;
+    struct evbuffer *buf = NULL;
+    int r = SQLITE_OK;
+
+    verifydb_get_rep rep = {};
+    rep.cb = sizeof(rep);
+    rep.status = VERIFYDB_ERROR;
+    rep.entry_size = sizeof(verifydb_entry);
+
+    r = sqlite3_prepare(Db, "SELECT timestamp, vmethod, vkey FROM verifydb "
+            "WHERE userid = lower(?) AND generation = ? AND "
+            "(1 = ? OR vmethod = ?) "
+            "ORDER BY vmethod ASC;",
+            -1, &Stmt, NULL);
+    if (r != SQLITE_OK)
+        goto end;
+
+    r = sqlite3_bind_text(Stmt, 1, req->userid, -1, SQLITE_STATIC);
+    if (r != SQLITE_OK)
+        goto end;
+    r = sqlite3_bind_int64(Stmt, 2, req->generation);
+    if (r != SQLITE_OK)
+        goto end;
+    r = sqlite3_bind_int64(Stmt, 3, req->vmethod == VMETHOD_UNSET ? 1 : 0);
+    if (r != SQLITE_OK)
+        goto end;
+    r = sqlite3_bind_int64(Stmt, 4, req->vmethod);
+    if (r != SQLITE_OK)
+        goto end;
+
+    buf = evbuffer_new();
+    if (!buf)
+        goto end;
+    while (sqlite3_step(Stmt) == SQLITE_ROW) {
+        const int64_t timestamp = sqlite3_column_int64(Stmt, 0);
+        const int32_t vmethod = sqlite3_column_int64(Stmt, 1);
+        const char *vkey = (const char *)sqlite3_column_text(Stmt, 2);
+        if (!vkey)
+            continue;  // Bad record.
```

Comment:
Can't you just filter these out with the WHERE clause?
This way you could also get rid of `evbuffer_prepend()` and
instead queue up everything in order, or even drop `evbuffer` altogether.

Reply:
My bad. It should return the record though. I made it return an empty string.
Right now, we won't have this case.

Reply:
There may be multiple records. It needs the count in the reply struct that is going to be prepended.